cbmc/proof_guide.md

```diff
@@ -2,57 +2,151 @@
 
 # CBMC Proof Guide and Cookbook for mlkem-native
 
-This doc acts as a guide to developing proofs of C code using [CBMC](https://model-checking.github.io/cbmc-training/). It concentrates
-on the use of _contracts_ to achieve _unbounded_ and _modular_ proofs of type-safety
-and correctness properties.
+This document acts as a guide to how we develop proofs of mlkem-native's C code using
+[CBMC](https://model-checking.github.io/cbmc-training/). It concentrates on the use of _contracts_ to achieve
+_unbounded_ and _modular_ proofs of type-safety and correctness properties.
 
-This document uses the abbreviated forms of the [CBMC contracts](https://diffblue.github.io/cbmc/contracts-user.html) defined by macros in the
-[cbmc.h](../mlkem/cbmc.h) header file in the mlkem-native sources.
+## Scope
 
-## Common Proof Patterns
+Our CBMC proofs confirm the absence of certain classes of undefined behaviour, such as integer overflow or out of bounds
+memory accesses -- for the precise list of conditions checked, see the CBMC configuration in
+[Makefile.common](Makefile.common). For many arithmetic functions, we additionally specify how they affect coefficient
+bounds: For example, we show that the result of `poly_invntt_tomont()` has coefficients bound by
+`INVNTT_BOUND`. Finally, some simple functions have their full functional behaviour specified: For example, the
+specification of the constant-time `ct_memcmp()` shows that, functionally, it is just an ordinary `memcmp()`.
 
-### Loops (general advice)
+## CBMC annotations
 
-1. A function should contain at most one outermost loop statement. If the function you're trying to prove has more than one outermost loop, then re-factor it into two or more functions.
+CBMC proofs are largely automatic; there are no proof scripts as in interactive theorem proving. Instead, CBMC consumes
+_annotated_ C code and checks that it does not exhibit the configured classes of undefined behaviour within the
+context described by the annotations; if the annotations add further constraints, those are checked, too. For example, a
+function contract annotation provides contextual assumptions about a function as _preconditions_ to CBMC, and adds
+further _constraints_ for the program state at function return.
 
-2. The one outermost loop statement should be the _final_ statement before the function returns. Don't have complicated code _after_ the loop body.
+In mlkem-native, we use abbreviated forms of the CBMC annotations defined by macros in the [cbmc.h](../mlkem/cbmc.h). We
+now list the most prominent.
+
+### Function contracts
+
+A _function contract_ can be added when a function is declared or defined; typically, non-`static` functions should have
```

Comment:
Are you sure you can define a contract in the function definition? 
Sounds like a foodgun because if in a proof using a certain function contract, you do not include the definition, it would see an empty contract from the declaration assuming no pre-conditions, while the proof of the contract would see the pre-conditions.

Should this be rephrased or am I misunderstanding something?

Reply:
This is mainly for `static` functions which often have definition and declaration at the same time. But you're right, it's a bit misleading, I'll rephrase.